### ISSUE TITLE
Using Bottom Up Feature

### ISSUE THREAD

**Issue by atg93** [husthuaan/AoANet#23]
Hi ,thank you for sharing your code. However, I encounter a problem. Currently, I cannot download bottom-up feature using your link  https://storage.googleapis.com/bottom-up-attention/trainval.zip. Moreover, I also used the link which is used in self-critical repo (https://imagecaption.blob.core.windows.net/imagecaption/trainval.zip), but downloading fails after a while. I was wondering is there another link that I can use ? By the way, bottom-up feature of flick30k data sets link also contains coco feature. Can I used them after converting to pth file ?

**Comment by husthuaan** [comment]
@atg93 
> I was wondering is there another link that I can use?

Yes, here is a bauduyun link (https://pan.baidu.com/s/142SqPM5-KEbzkV2KrhZdWQ) which contains identical files of the original link. And you can try another google link (https://drive.google.com/file/d/1MV6dSnqViQfyvgyHrmAT_lLpFbkzp3mx/view) from https://github.com/aimagelab/meshed-memory-transformer, but the file format is different.
> By the way, bottom-up feature of flick30k data sets link also contains coco feature. Can I used them after converting to pth file ?

I guess yes. But the format may be different (I am not sure).

**Comment by atg93** [comment]
Thank you very much for your answer. By the way, is it also possible that you could share generated captions from Coco and flickr30k test datasets ? It will be very helpful.

**Comment by husthuaan** [comment]
Hi @atg93, I have uploaded the generated captions for COCO test and val splits at [here](https://drive.google.com/drive/folders/1jAPRkhK90VJj-fqLrdsVS5wJhRRMwShf). However, we don't have the generated captions for flickr30k, because we didn't experiment on this dataset.

I am sorry about the delay.

**Comment by sssilence** [comment]
Excuse me. I have downloaded trainval.But when I perform python scripts/make_bu_data.py --output_dir data/cocobu, I encountered this error“_csv.Error: iterator should return strings, not bytes (did you open the file in text mode?)”.And I have tried some modifications，but I haven't solved it.Can you give me some guidance？Thanks!!! @husthuaan @atg93

**Comment by Huntersxsx** [comment]
> Excuse me. I have downloaded trainval.But when I perform python scripts/make_bu_data.py --output_dir data/cocobu, I encountered this error“_csv.Error: iterator should return strings, not bytes (did you open the file in text mode?)”.And I have tried some modifications，but I haven't solved it.Can you give me some guidance？Thanks!!! @husthuaan @atg93

@sssilence 
Hi, I encountered the same error with you, and I replaced 'r+b' with 'rt', and "item[field]" => "bytes(item[field], 'utf-8')", but I failed in line 47 "np.savez_compressed", with the error "SyntaxError: invalid syntax". Do you have any idea?

**Comment by Huntersxsx** [comment]
> > Excuse me. I have downloaded trainval.But when I perform python scripts/make_bu_data.py --output_dir data/cocobu, I encountered this error“_csv.Error: iterator should return strings, not bytes (did you open the file in text mode?)”.And I have tried some modifications，but I haven't solved it.Can you give me some guidance？Thanks!!! @husthuaan @atg93
> 
> @sssilence
> Hi, I encountered the same error with you, and I replaced 'r+b' with 'rt', and "item[field]" => "bytes(item[field], 'utf-8')", but I failed in line 47 "np.savez_compressed", with the error "SyntaxError: invalid syntax". Do you have any idea?

@sssilence 
Sorry to bother you, I just missed ')' 😂
To solve your problem, you can
change the line 40:
with open(os.path.join(args.downloaded_feats, infile), "r+b") as tsv_in_file:
to:
with open(os.path.join(args.downloaded_feats, infile), "rt") as tsv_in_file:
and change the line 44:
item[field] = np.frombuffer(base64.decodestring(item[field]),
to
item[field] = np.frombuffer(base64.decodestring(bytes(item[field],'utf-8')),
Hope it can help you

**Comment by sssilence** [comment]
> > > Excuse me. I have downloaded trainval.But when I perform python scripts/make_bu_data.py --output_dir data/cocobu, I encountered this error“_csv.Error: iterator should return strings, not bytes (did you open the file in text mode?)”.And I have tried some modifications，but I haven't solved it.Can you give me some guidance？Thanks!!! @husthuaan @atg93
> > 
> > 
> > @sssilence
> > Hi, I encountered the same error with you, and I replaced 'r+b' with 'rt', and "item[field]" => "bytes(item[field], 'utf-8')", but I failed in line 47 "np.savez_compressed", with the error "SyntaxError: invalid syntax". Do you have any idea?
> 
> @sssilence
> Sorry to bother you, I just missed ')' 😂
> To solve your problem, you can
> change the line 40:
> with open(os.path.join(args.downloaded_feats, infile), "r+b") as tsv_in_file:
> to:
> with open(os.path.join(args.downloaded_feats, infile), "rt") as tsv_in_file:
> and change the line 44:
> item[field] = np.frombuffer(base64.decodestring(item[field]),
> to
> item[field] = np.frombuffer(base64.decodestring(bytes(item[field],'utf-8')),
> Hope it can help you

Thank you very much!!!Your method is very effective.I have solved the error.Thank you again.

**Comment by aLefred0** [comment]
> @atg93
> 
> > I was wondering is there another link that I can use?
> 
> Yes, here is a baiduyun link (https://pan.baidu.com/s/142SqPM5-KEbzkV2KrhZdWQ) which contains identical files of the original link. And you can try another google link (https://drive.google.com/file/d/1MV6dSnqViQfyvgyHrmAT_lLpFbkzp3mx/view) from https://github.com/aimagelab/meshed-memory-transformer, but the file format is different.
> 
> > By the way, bottom-up feature of flick30k data sets link also contains coco feature. Can I used them after converting to pth file ?
> 
> I guess yes. But the format may be different (I am not sure).

Hi. I was facing similar issues where the download fails after sometime. I am facing a similar issue for the new links also. 
Do you have any alternate sources?

**Comment by aLefred0** [comment]
> > @atg93
> > > I was wondering is there another link that I can use?
> > 
> > 
> > Yes, here is a baiduyun link (https://pan.baidu.com/s/142SqPM5-KEbzkV2KrhZdWQ) which contains identical files of the original link. And you can try another google link (https://drive.google.com/file/d/1MV6dSnqViQfyvgyHrmAT_lLpFbkzp3mx/view) from https://github.com/aimagelab/meshed-memory-transformer, but the file format is different.
> > > By the way, bottom-up feature of flick30k data sets link also contains coco feature. Can I used them after converting to pth file ?
> > 
> > 
> > I guess yes. But the format may be different (I am not sure).
> 
> Hi. I was facing similar issues where the download fails after sometime. I am facing a similar issue for the new links also.
> Do you have any alternate sources?
Update:I got the features from the old link

**Comment by A1ist** [comment]
> > > @atg93
> > > > I was wondering is there another link that I can use?
> > > 
> > > 
> > > Yes, here is a baiduyun link (https://pan.baidu.com/s/142SqPM5-KEbzkV2KrhZdWQ) which contains identical files of the original link. And you can try another google link (https://drive.google.com/file/d/1MV6dSnqViQfyvgyHrmAT_lLpFbkzp3mx/view) from https://github.com/aimagelab/meshed-memory-transformer, but the file format is different.
> > > > By the way, bottom-up feature of flick30k data sets link also contains coco feature. Can I used them after converting to pth file ?
> > > 
> > > 
> > > I guess yes. But the format may be different (I am not sure).
> > 
> > 
> > Hi. I was facing similar issues where the download fails after sometime. I am facing a similar issue for the new links also.
> > Do you have any alternate sources?
> > Update:I got the features from the old link

Hi, I am also facing the problem of failure to download for a period of time and have tried for many days with the old link without success. Can you share me a new link if you succeed in downloading? Thank you very much!
